engine/consensus/matching/engine.go

```diff
@@ -322,6 +325,21 @@ func (e *Engine) onReceipt(originID flow.Identifier, receipt *flow.ExecutionRece
 	return nil
 }
 
+func (e *Engine) persistExecutionReceipt(receipt *flow.ExecutionReceipt) error {
+	// store the receipt to make it persistent for later
+	err := e.receiptsDB.Store(receipt) // internally de-duplicates
+	if err != nil && !errors.Is(err, storage.ErrAlreadyExists) {
+		return fmt.Errorf("could not store receipt: %w", err)
+	}
+	// TODO if the second operation fails we should remove stored execution result
+	// This is global execution storage problem - see TODO at the top
+	err = e.receiptsDB.Index(receipt.ExecutionResult.BlockID, receipt.ID())
```

Comment:
I think we should remove the `Index` method, and let `Store` instead it internally within the same transaction.

Otherwise, there will be more edge cases about data inconsistency. For instance, receipt is stored, but indexing failed.

Reply:
It's already that way for execution engine so I've just copied it, I can remove `Index` from interface though